src/Enzo/enzo_EnzoMethodFeedbackSTARSS.cpp

```diff
@@ -0,0 +1,1489 @@
+
+#define DEBUG_FEEDBACK_STARSS
+//#define DEBUG_FEEDBACK_STARSS_SN
+
+// =============================================================================
+// splice these off to a different file (later)
+// TODO: Maybe create EnzoStarParticle class and add rate-calculating functions there?
+
+int EnzoMethodFeedbackSTARSS::determineSN(double age, int* nSNII, int* nSNIA,
+                double mass_Msun, double tunit, float dt){
+
+    const EnzoConfig * enzo_config = enzo::config();
+    std::mt19937 mt(std::time(nullptr));
+
+    if (NEvents > 0){
+        *nSNII = 1;
+        NEvents -= 1;
+        return 1;
+    }
+    /* else, calculate SN rate, probability and determine number of events */
+    *nSNII = 0;
+    *nSNIA = 0;
+    double RII=0, RIA=0, PII=0, PIA=0, random = 0;
+    if (enzo_config->method_feedback_single_sn && NEvents < 0)
+    {
+        /* age-dependent rates */
+        if (age < 3.401)
+        {
+            RII = 0.0;
+            RIA = 0.0;
+        }
+        if (3.401 <= age && age< 10.37)
+        {
+            RII = 5.408e-4;
+            RIA = 0.0;
+        }
+        if (10.37 <= age && age < 37.53)
+        {
+            RII = 2.516e-4;
+            RIA = 0.0;
+        }
+        if (37.53 <= age)
+        {
+            RII = 0.0;
+            RIA = 5.3e-8+1.6e-5*exp(-0.5*pow((age-50.0)/10.0, 2));
+        }
+        /* rates -> probabilities */
+        if (RII > 0){
+            PII = RII * mass_Msun / cello::Myr_s *tunit*dt;
+            double random = double(mt())/double(mt.max());
+            if (PII > 1.0 && enzo_config->method_feedback_unrestricted_sn){
+                int round = (int)PII;
+                *nSNII = round;
+                PII -= round;
+            }
+            if (PII > 1.0 && !enzo_config->method_feedback_unrestricted_sn){
+                //if it wants to do > 1 SN per particle, but unrestricted_sn=false,
+                //only do 1 SN
+
+                PII = 1.0;
+                *nSNII = 0;
+            }
+            int psn = *nSNII;
+            if (random <= PII){
+                *nSNII = psn+1;
+            }
+        }
+        #ifdef DEBUG_FEEDBACK_STARSS_SN
+          CkPrintf("MethodFeedbackSTARSS::determineSN() -- mass_Msun = %f; age = %f; RII = %f; RIA = %f\n",
+                    mass_Msun, age, RII, RIA);
+        #endif
+
+        if (RIA > 0){
+            PIA = RIA*mass_Msun / cello::Myr_s *tunit*dt;
+            float random = float(rand())/float(RAND_MAX);
+
+            if (PIA > 1.0 && enzo_config->method_feedback_unrestricted_sn)
+            {
+                int round = int(PIA);
+                *nSNIA = round;
+                PIA -= round;
+            }
+            int psn = *nSNIA;
+
+            if (random < PIA)
+                *nSNIA = psn+1;
+        }
+        #ifdef DEBUG_FEEDBACK_STARSS_SN
+          CkPrintf("MethodFeedbackSTARSS::determineSN() -- PII = %f; PIA = %f\n", PII, PIA);
+        #endif
+    }
```

Comment:
`RII` and `RIA` are supernova rates in units of `1/Mass/Time`. Multiplying by `mass_Msun * dt` should give you the number of SNe that you would expect to go off in a timestep given these rates. Variable name left over from copy/paste, I'll update it to `number_per_dt` or something like that.

If `method_feedback_unrestriced_sn` is false, then this will limit the number of SNe to 1 for this particle.

Reply:
Added clarifying comments and enforced that for `unrestricted_sn=false`, it will only set off a Type IA supernova if it's been determined that no Type II supernovae will go off. How it was written before, no Type IA supernovae would go off if `unrestricted_sn=false`, which is not the intended behavior.